src/js/common/validator.js

```diff
@@ -1,122 +1,540 @@
+     *          dataPath: ["deep", "booleanField"],
+     *          schemaPath: ["properties", "deep", "properties", "booleanField", "type"],
+     *          rule: { type: "boolean" },
+     *          message: "validation-invalid-field-type"
+     *      }
+     *  ]
+     * }
+     *
+     * Note that as in the examples above, the GSS metaschema defines "error hints" that result in message keys
+     * being returned.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {Object} gssSchema - A GSS schema.
+     * @param {ajvErrors|Boolean} ajvErrors - The raw errors returned by AJV, if there are any, or `false` if there are no validation Errors.
+     * @return {Object} - An object detailing the validation results (see above).
+     *
+     */
+    gpii.schema.validator.standardiseAjvErrors = function (gssSchema, ajvErrors) {
+        if (ajvErrors) {
+            var transformedErrors = fluid.transform(ajvErrors, function (ajvError) {
+                var error         = {};
+
+                error.dataPath = gpii.schema.validator.extractElDataPathSegmentsFromError(ajvError);
+
+                var schemaPath = gpii.schema.validator.extractElSchemaPathSegmentsFromError(ajvError);
+                error.schemaPath = schemaPath;
+
+                // Backtrack from the full path one step so that we get a bit more context, i.e. `{ "type": "string" }` instead of `string`.
+                var fullRulePath = schemaPath.slice(0, -1);
+                error.rule = fluid.get(gssSchema, fullRulePath);
+
+                // Use any error hints found for the failing path in the schema, failing over to the AJV message.
+                error.message = gpii.schema.validator.errorHintForRule(schemaPath, gssSchema, ajvError.message);
+
+                return error;
+            });
+            return {
+                isValid: false,
+                errors: transformedErrors
+            };
+        }
+        else {
+            return { isValid: true, errors: [] };
+        }
+    };
+
+    fluid.registerNamespace("gpii.schema.validator");
+    gpii.schema.validator.defaultLocalisationTransformRules = {
+        "data": "data",
+        "error": "error"
+    };
+
+    /**
+     *
+     * A function to translate/localise validation errors.
+     *
+     * If you want to pass a custom message bundle to this function, it should only contain top-level elements, see
+     * ./src/js/validation-errors.js in this package for an example.
+     *
+     * @param {Array<Object>} validationErrors - An array of validation errors, see `gpii.schema.validator.standardiseAjvErrors` for details.
+     * @param {Any} validatedData - The (optional) data that was validated.
+     * @param {Object} messages - An (optional) map of message templates (see above).  Defaults to the message bundle provided by this package.
```

Comment:
I think we can help here with some compound type such as {Object.<String, String>} ?

Reply:
Yup, done.